actors/miner/src/lib.rs

```diff
@@ -2995,6 +2996,301 @@ impl Actor {
+                                illegal_argument,
+                                "invalid post was submitted"
+                            )
+                            .into());
+                        }
+                        Ok(())
+                    })
+                    .map_err(|e| {
+                        e.downcast_default(ExitCode::USR_ILLEGAL_STATE, "while removing partitions")
+                    })?;
+            }
+
+            if !deadline_available_for_compaction(
+                policy,
+                current_deadline.period_start,
+                params.to_deadline,
+                rt.curr_epoch(),
+            ) {
+                return Err(actor_error!(
+                    forbidden,
+                    "cannot move to deadline {} during its challenge window, \
+                        or the prior challenge window,
+                        or before {} epochs have passed since its last challenge window ended",
+                    params.to_deadline,
+                    policy.wpost_dispute_window
+                ));
+            }
+
+            if !deadline_available_for_move(
+                policy,
+                params.from_deadline,
+                params.to_deadline,
+                current_deadline.index,
+            ) {
+                return Err(actor_error!(
+                    forbidden,
+                    "cannot move to a deadline which is further from current deadline"
+                ));
+            }
+
+            let from_quant = state.quant_spec_for_deadline(policy, params.from_deadline);
+            let to_quant = state.quant_spec_for_deadline(policy, params.to_deadline);
+
+            let mut from_deadline =
+                deadlines.load_deadline(policy, store, params.from_deadline).map_err(|e| {
+                    e.downcast_default(
+                        ExitCode::USR_ILLEGAL_STATE,
+                        format!("failed to load deadline {}", params.from_deadline),
+                    )
+                })?;
+            let mut to_deadline =
+                deadlines.load_deadline(policy, store, params.to_deadline).map_err(|e| {
+                    e.downcast_default(
+                        ExitCode::USR_ILLEGAL_STATE,
+                        format!("failed to load deadline {}", params.to_deadline),
+                    )
+                })?;
+
+            let partitions = &mut params.partitions;
+            if partitions.is_empty() {
+                let partitions_amt = from_deadline.partitions_amt(rt.store()).map_err(|e| {
+                    e.downcast_default(
+                        ExitCode::USR_ILLEGAL_STATE,
+                        format!("failed to load partitions for deadline {}", params.from_deadline),
+                    )
+                })?;
+
+                for partition_idx in 0..partitions_amt.count() {
+                    partitions.set(partition_idx);
+                }
```

Comment:
@anorth it's a convenient default behavior that when no partitions is specified, moving all partitions from the source deadline.

Reply:
Oh I see. I don't have a super-strong objection, but I don't think its worthwhile.
- This will not be a common operation, so the optimisation isn't high impact
- Since partitions are specified as a RLE bitfield, "all partitions" is only a single integer on-chain anyway

Given that, I think requiring explicit identification of partitions is better. An SP who empties a deadline by accident could cause themselves trouble, and we should make that accident a bit harder.

Reply:
Make sense, empty partitions are now forbidden.